app/models/line_item.rb

```diff
@@ -2,6 +2,7 @@ class LineItem < ApplicationRecord
   belongs_to :order, optional: true
   belongs_to :product
   belongs_to :cart, optional: true
+  validates :product, uniqueness: {scope: :cart}
```

Comment:
is there any difference if we used product_id and cart_id instead of product and cart here
Please read about allow_blank and allow_nil and tell me the difference and see if it can/should be used here

Reply:
`validates :product, uniqueness: {scope: :cart}` will also look for product_id and cart_id but its better to pass the attribute rather than whole object so we should use product_id and cart_id here.

`allow_nil` skips validation for nil values
`allow_blank` skips validation for nil and empty string values 
default value for both is 'false'

Here we don't need it because when we insert a nil or blank value then `belongs_to :product` will raise an error.